### PR TITLE
docs: Convert examples within the imagebufalgo chapter.

### PR DESCRIPTION
**Pull request by buddly27**
## Description
Convert C++ and Python examples from the "imagebufalgo" chapter into tests within the "docs-examples" testsuites (#3992).

## Checklist:

- [x] I have read the [contribution guidelines](https://github.com/AcademySoftwareFoundation/OpenImageIO/blob/master/CONTRIBUTING.md).
- [x] I have updated the documentation, if applicable.
- [x] I have ensured that the change is tested somewhere in the testsuite
  (adding new test cases if necessary).
- [x] If I added or modified a C++ API call, I have also amended the
  corresponding Python bindings (and if altering ImageBufAlgo functions, also
  exposed the new functionality as oiiotool options).
- [x] My code follows the prevailing code style of this project. If I haven't
  already run clang-format before submitting, I definitely will look at the CI
  test that runs clang-format and fix anything that it highlights as being
  nonconforming.

### CODE REVIEW

File: src/doc/imagebufalgo.rst [hunk 1]
```diff
@@ -1075,7 +1071,7 @@ Shuffling channels
 
     .. tabs::
   
-       .. code-tab:: c++
+       .. code-tab:: c++`
```
Comment (MichaelRCarroll-Intel) [inline src/doc/imagebufalgo.rst:1074]: Is this backtick a typo?... or perhaps there is something I'm not seeing something about needed syntax in .rst?

Thanks for the diff... Seeing everything should make mine quicker.

Comment (buddly27) [reply src/doc/imagebufalgo.rst:1074]: Definitely a typo, thanks for checking, I'll remove it on the next pass!

File: testsuite/docs-examples-cpp/ref/out.txt [hunk 2]
```diff
@@ -1,3 +1,15 @@
+error: Uninitialized input image
+error: Uninitialized input image
```
Comment (lgritz) [inline testsuite/docs-examples-cpp/ref/out.txt:2]: I'm a little concerned about these lines... is something in an example not working?

File: testsuite/docs-examples-cpp/run.py [hunk 3]
```diff
@@ -11,7 +11,7 @@
     prefix = "./build/"
 
 # Prep:
-command += run_app("cmake -E copy ../common/grid-small.exr grid.exr")
+command += run_app("cmake -E copy ${OIIO_TESTSUITE_ROOT}/common/grid-small.exr grid.exr")
```
Comment (lgritz) [inline testsuite/docs-examples-cpp/run.py:14]: Why this change?

File: testsuite/docs-examples-cpp/ref/out.txt (hunk 2 above)
Comment (lgritz) [reply testsuite/docs-examples-cpp/ref/out.txt:2]: Oh, never mind, those are from the ones where you are purposely making an error and showing how to check it, right?

File: testsuite/docs-examples-python/run.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 
 
 # Prep:
-command += run_app("cmake -E copy ../common/grid-small.exr grid.exr")
+command += run_app("cmake -E copy ${OIIO_TESTSUITE_ROOT}/common/grid-small.exr grid.exr")
```
Comment (lgritz) [inline testsuite/docs-examples-python/run.py:9]: Did something not work before?

I'm not sure this is what you want here. The test will be run from within `build/testsuite/<testname>`, and it really is ../common where the grid-small.exr is located.

I believe OIIO_TESTSUITE_ROOT is the *source* testsuite, not the copy in build, if you know what I mean.

**Comment by lgritz** [comment]
Other than the comment about OIIO_TESTSUITE_ROOT, this is all looking really good! You didn't get tripped by anything I might have expected. Excellent work!

File: testsuite/docs-examples-cpp/ref/out.txt (hunk 2 above)
Comment (buddly27) [reply testsuite/docs-examples-cpp/ref/out.txt:2]: This is the result of testing the error outputs:
https://openimageio.readthedocs.io/en/v2.5.4.0/imagebufalgo.html#return-values-and-error-messages

I reproduced the tests, which print out "error: ..." for the two first examples:
https://github.com/buddly27/OpenImageIO/blob/pr3992/testsuite/docs-examples-cpp/src/docs-examples-imagebufalgo.cpp#L46
https://github.com/buddly27/OpenImageIO/blob/pr3992/testsuite/docs-examples-cpp/src/docs-examples-imagebufalgo.cpp#L59

This could potentially be tested differently:
```cpp
// BEGIN-imagebufalgo-output-error2
// Method 2: Write into an existing image
ImageBuf dst;   // will be the output image
bool ok = ImageBufAlgo::over (dst, fg, bg);
std::string error;
if (! ok)
    error = dst.geterror();
// END-imagebufalgo-output-error2

static_assert(!error.empty(), "An error should have been set");
```

Would you prefer that?

File: testsuite/docs-examples-python/run.py (hunk 4 above)
Comment (buddly27) [reply testsuite/docs-examples-python/run.py:9]: It was because the relative path resolution was failing when running "ctest" as follows:
```console
ctest --test-dir ./build 
```

I just now realized you provided a way to run single tests with `make test TEST=...`, I thought using an absolute path would offer more flexibility. Also, being in "source" is not necessarily a blocker, as the copied image will be there:
https://github.com/buddly27/OpenImageIO/blob/pr3992/testsuite/common/grid-small.exr

We could also use a CMAKE variable (e.g. [CMAKE_CURRENT_SOURCE_DIR](https://cmake.org/cmake/help/latest/variable/CMAKE_CURRENT_SOURCE_DIR.html)) but I'm not entirely sure if it would be resolved as expected within the python scripts.

File: testsuite/docs-examples-cpp/ref/out.txt (hunk 2 above)
Comment (lgritz) [reply testsuite/docs-examples-cpp/ref/out.txt:2]: No, I think it's fine as it is. I just saw the error message in the captured output and momentarily forgot that this was expected.

File: testsuite/docs-examples-python/run.py (hunk 4 above)
Comment (lgritz) [reply testsuite/docs-examples-python/run.py:9]: The case I'm actually concerned about is that sometimes (willing to entertain the notion that it's nonstandard and not necessarily a good idea) I do this to run a test by hand, completely outside of ctest/cmake/make, usually when I'm tracking down some kind of broken behavior:

    cd build/testsuite/foo
    python3 ../runtest.py

But I'm not overly worried about this. I can probably find another way to do the kind of debugging I do this way.

Comment (lgritz) [reply testsuite/docs-examples-python/run.py:9]: But my point is, if you execute the test script by hand outside of ctest, the OIIO_TESTSUITE_ROOT environment variable will not be set!

Comment (buddly27) [reply testsuite/docs-examples-python/run.py:9]: Oh, I see, this will be a problem indeed. Then what about using the "test_source_dir" variable that is defined within `runtest.py`?
```python
command += run_app("cmake -E copy " + test_source_dir + "/../common/grid-small.exr grid.exr")
```

Comment (lgritz) [reply testsuite/docs-examples-python/run.py:9]: sure, I think that's fine

Comment (buddly27) [reply testsuite/docs-examples-python/run.py]: Updated!

File: testsuite/docs-examples-cpp/run.py (hunk 3 above)
Comment (buddly27) [reply testsuite/docs-examples-cpp/run.py]: Related to other [note](https://github.com/AcademySoftwareFoundation/OpenImageIO/pull/4016#discussion_r1357463066)

**Review by lgritz — APPROVED** [review]
LGTM, this is great!
